common/channel/src/libra_channel.rs

```diff
@@ -0,0 +1,151 @@
+use std::collections::{HashMap, VecDeque};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::Waker;
+
+use futures::async_await::FusedStream;
+use futures::stream::Stream;
+use futures::task::Context;
+use futures::Poll;
+
+/// ValidatorMessage trait needs to be implemented by every message which gets pushed into the Libra
+/// channel so that we can ensure fairness among validators
+pub trait ValidatorMessage {
+    /// Extract the Validator from which this message arrived. This
+    /// will be used for ensuring fairness among validators.
+    fn get_validator(&self) -> &str;
+}
+
+/// Type is an enum which can be used as a configuration option for
+/// libra_channel
+pub enum Type {
+    LIFO,
+    FIFO,
+}
+
+struct PerValidatorQueue<T> {
+    queue_type: Type,
+    per_validator_queue: HashMap<String, VecDeque<T>>,
+    max_queue_size: usize,
+}
+
+impl<T: ValidatorMessage> PerValidatorQueue<T> {
+    fn new(queue_type: Type, max_queue_size: usize) -> Self {
+        assert!(
+            max_queue_size > 0,
+            "cannot have a zero-capacity libra_channel"
+        );
+        Self {
+            queue_type,
+            per_validator_queue: HashMap::new(),
+            max_queue_size,
+        }
+    }
+
+    fn add_message(&mut self, message: T) {
```

Comment:
nit: Could you please add a high-level comment describing the the overall structure and the semantics? (just a couple of lines)

Reply:
Added